lib/postgrex/extensions/inet.ex

```diff
@@ -19,9 +19,11 @@ defmodule Postgrex.Extensions.INET do
   def decode(_) do
     quote location: :keep do
       <<8 :: int32, 2, n, _cidr?, 4, a, b, c, d>> ->
+        n = if(n == 32, do: nil, else: n)
         %Postgrex.INET{address: {a, b, c, d}, netmask: n}
       <<20 :: int32, 3, n, _cidr?, 16,
         a::16, b::16, c::16, d::16, e::16, f::16, g::16, h::16>> ->
+          n = if(n == 128, do: nil, else: n)
```

Comment:
Can we always set the netmask as an integer on decode but handle `nil` on the encoding side (as is done already). Otherwise we lose the netmask information. Following the generous with what we allow as input but strict with what we have as output.

Reply:
I believe that the correct behaviour is what I have written here, and matches what is displayed from `psql`:

```sql
CREATE TABLE IF NOT EXISTS inet_example (id bigserial, address inet);
INSERT INTO inet_example (address) VALUES
('127.0.0.1'::inet), ('127.0.0.1/32'::inet), ('127.0.0.1/16'::inet),
('::1'::inet), ('::1/128'::inet), ('::1/16'::inet), ('::1/32'::inet);
SELECT address FROM inet_example;
```

The output is:
```
┌──────────────┐
│   address    │
├──────────────┤
│ 127.0.0.1    │
│ 127.0.0.1    │
│ 127.0.0.1/16 │
│ ::1          │
│ ::1          │
│ ::1/16       │
│ ::1/32       │
└──────────────┘
```

This is exactly the conversion that I am doing on decode, because it’s what most people would expect to be the case for an IP address. When you’re asked for your IP address, you typically answer `10.10.2.1`, not `10.10.2.1/32` (which a network admin will say is correct). From 8.9.1 inet in the PostgreSQL manual:

> The `inet` type holds an IPv4 or IPv6 host address, and optionally its subnet, all in one field. The subnet is represented by the number of network address bits present in the host address (the “netmask”). If the netmask is 32 and the address is IPv4, then the value does not indicate a subnet, only a single host. In IPv6, the address length is 128 bits, so 128 bits specify a unique host address. Note that if you want to accept only networks, you should use the `cidr` type rather than `inet`.
> The input format for this type is *`address/y`* where address is an IPv4 or IPv6 address and y is the number of bits in the netmask. If the *`/y`* portion is missing, the netmask is 32 for IPv4 and 128 for IPv6, so the value represents just a single host. On display, the *`/y`* portion is suppressed if the netmask specifies a single host.

This might be the distinction that I want to put for adam12/ecto_network#11 so that there’s a different interpretation in the human representation for Ecto (that is, it would always show `127.0.0.1/32` if the type is CIDR).

There’s an important *logical* distinction between INET and CIDR. `SELECT cidr '127.0.0.1/32';` is legal, but `SELECT cidr '127.0.0.1/24'` is not because CIDR does not allow nonzero bits to the right of the netmask, which can result in errors. There does appear to be automatic safe conversion because `SELECT cidr (inet '127.0.0.1/24')` returns the expected `127.0.0.0/24`, so I’m ultimately not that concerned about that logical distinction (and I tested by providing an `inet` to a column expecting a `cidr`).

Reply:
I should point out, after reading your second comment, @fishcakez, that when I `SELECT address::text FROM inet_example`, this is the result:

```
┌──────────────┐
│   address    │
├──────────────┤
│ 127.0.0.1/32 │
│ 127.0.0.1/32 │
│ 127.0.0.1/16 │
│ ::1/128      │
│ ::1/128      │
│ ::1/16       │
│ ::1/32       │
└──────────────┘
```

The netmask is displayed. But this is not the default display from psql.

Reply:
That sounds good. It will probably take a day or two to do this, including the test changes requested.

Reply:
No rush, it's better we get this right.